compiler/rustc_span/src/hygiene.rs

```diff
@@ -1405,23 +1437,37 @@ pub fn decode_syntax_context<D: Decoder, F: FnOnce(&mut D, u32) -> SyntaxContext
         }
     };
 
-    // Don't try to decode data while holding the lock, since we need to
-    // be able to recursively decode a SyntaxContext
-    let mut ctxt_data = decode_data(d, raw_id);
-    // Reset `dollar_crate_name` so that it will be updated by `update_dollar_crate_names`
-    // We don't care what the encoding crate set this to - we want to resolve it
-    // from the perspective of the current compilation session
-    ctxt_data.dollar_crate_name = kw::DollarCrate;
+    let ctxt_key = decode_ctxt_key(d, raw_id);
 
-    // Overwrite the dummy data with our decoded SyntaxContextData
-    HygieneData::with(|hygiene_data| {
-        if let Some(old) = hygiene_data.syntax_context_data.get(raw_id as usize)
```

Comment:
> That's because all hygiene logic runs on comparing SyntaxContext, not SyntaxContextDatas, so two different SyntaxContexts are considered different even if they datas are the same.

I think this is exactly what #127279 aims to do: reuse the `opaque` value to ensure it compares the correct `SyntaxContext` value.

> If two syntax contexts have matching (parent, outer_expn_id, outer_transparency) tuples, they shouldn't just get identical (cloned) SyntaxContextDatas, they should have the same SyntaxContext ids.

I'm confused because the old code decodes `opaque` and generates a new `SyntaxContext` value using [new_ctxt](https://github.com/rust-lang/rust/pull/129827/files#diff-1f9c21b64fad30a12ac049ab77bfd50ab32a8b3e666bef801ebb906698bd1cfcR1421). Following this view, this logic has always been flawed.

Reply:
>Following this view, this logic has always been flawed.

Yep, that seems true.
I'm still investigating.